src/jubjub/mod.rs

```diff
@@ -434,6 +460,260 @@ impl JubjubBls12 {
+    let n_segments_as_bytes = (n_segments as u32).to_le_bytes();
+    let window_size_as_bytes = window_size.to_le_bytes();
+    file.write_all(&n_segments_as_bytes).expect("failed to write n_segments");
+    file.write_all(&window_size_as_bytes).expect("failed to write window_size");
+
+    let mut n_segments_written = 0;
+
+    // Generate and write the exp-table in chunks.
+    while n_segments_written < n_segments {
+        let n_segments_remaining = n_segments - n_segments_written;
+
+        // The number of segments of the exp-table being generated in this chunk. It is possible for
+        // the last chunk to contain fewer than `MAX_EXP_TABLE_SEGMENTS_IN_MEMORY` number of
+        // segments of the exp-table.
+        let n_segments_in_range = if n_segments_remaining > MAX_EXP_TABLE_SEGMENTS_IN_MEMORY {
+            MAX_EXP_TABLE_SEGMENTS_IN_MEMORY
+        } else {
+            n_segments_remaining
+        };
+
+        // Calculate the range of segment indices for this chunk, create the generators
+        // corresponding to each segment index, then create the exp-table corresponding to each of
+        // the generators.
+        let first_segment = n_segments_written;
+        let stop_at_segment = first_segment + n_segments_in_range;
+        let generators = &generators[first_segment..stop_at_segment];
+        let exp_table = create_exp_table_for_generators(generators, window_size);
+
+        // For each segment in this chunk, compress the segment's exp-table values (Edward points)
+        // in parallel. Collect into a vector then write to the exp-table file to reduce the number
+        // of file I/O operations.
+        for segment_table in exp_table {
+            let segment_table_bytes: Vec<u8> = segment_table
+                .par_iter()
+                .flat_map(|window| {
+                    window
+                        .par_iter()
+                        .flat_map(|point| point.compress().as_bytes())
+                        .collect::<Vec<u8>>()
+                })
+                .collect();
+
+            file.write_all(&segment_table_bytes).expect("failed to write segment table bytes");
+        }
+
+        n_segments_written += n_segments_in_range;
+    }
+}
+
+/// Loads into memory a range of the exp-table file. The range of the exp-table file that is read is
+/// given by the range of segment indices: `first_segment..first_segment + n_segments`.
+///
+/// This function can return less than `n_segments` worth of the exp-table if `first_segment +
+/// n_segments` exceeds the length of the exp-table file. If `first_segment` exceeds the number of
+/// segments in the exp-table file an empty vector is returned. It is the user's responsibility to
+/// check the number of exp-table segments returned from this function.
+pub fn read_exp_table_range(
+    first_segment: usize,
+    n_segments: usize,
+    exp_table_path: &str,
+) -> Vec<Vec<Vec<edwards::Point<Bls12, PrimeOrder>>>> {
```

Comment:
it is probably better if we return a `io::Result` here, so that there is some way of handling failed file reads/writes